File: app/core/combat/dice_resolver.py
from __future__ import annotations

import random
from enum import Enum

from app.core.constants import (
    BOOSTED_MULTIPLIER,
    EVASION_COUNTER_BASE,
    EVASION_COUNTER_MIN,
    EVASION_COUNTER_STEP,
    GUARD_UP_MITIGATION,
    TUMBLING_DELAY_MAX,
    TUMBLING_DELAY_MIN,
    TUMBLING_MULTIPLIER,
)


class DiceOutcome(Enum):
    BOOSTED  = "Boosted"
    SUCCESS  = "Success"
    TUMBLING = "Tumbling"
    GUARD_UP = "GuardUp"
    EVASION  = "Evasion"


class DiceResolver:
    """Resolves a single dice roll against a shifted probability table."""

    def roll(self, shifted_probabilities: dict) -> DiceOutcome:
        """
        Pick an outcome by weighted random selection.
        shifted_probabilities must sum to 1.0.
        """
        outcomes = list(shifted_probabilities.keys())
        weights  = [shifted_probabilities[k] for k in outcomes]
        chosen   = random.choices(outcomes, weights=weights, k=1)[0]
        return DiceOutcome(chosen)

    def apply_outcome(self, outcome: DiceOutcome, raw_output: int) -> dict:
        """
        Apply an outcome to raw_output and return a result dict:
          output         — final damage / healing value
          tumble_delay   — ticks added to attacker's timeline (Tumbling only)
          guard_mitigation — HP added to target's mitigation pool (GuardUp only)
          evaded         — True when the attack was fully evaded
        """
        result = {
            "output":           0,
            "tumble_delay":     0,
            "guard_mitigation": 0,
            "evaded":           False,
        }

        if outcome == DiceOutcome.BOOSTED:
            result["output"] = round(raw_output * BOOSTED_MULTIPLIER)

        elif outcome == DiceOutcome.SUCCESS:
            result["output"] = raw_output

        elif outcome == DiceOutcome.TUMBLING:
            result["output"]      = round(raw_output * TUMBLING_MULTIPLIER)
            result["tumble_delay"] = self.calculate_tumbling_delay()

        elif outcome == DiceOutcome.GUARD_UP:
            result["output"]           = raw_output
            result["guard_mitigation"] = round(raw_output * GUARD_UP_MITIGATION)

        elif outcome == DiceOutcome.EVASION:
            result["evaded"] = True

        return result

    def calculate_tumbling_delay(self) -> int:
        """Return a random tick delay within the Tumbling range."""
        return random.randint(TUMBLING_DELAY_MIN, TUMBLING_DELAY_MAX)

    def resolve_evasion_counter(self, recursion_depth: int) -> bool:
        """
        Attempt an evasion counter at the given recursion depth.
        Counter chance diminishes by EVASION_COUNTER_STEP per depth,
        floored at EVASION_COUNTER_MIN.
        Returns True if the counter triggers (attacker retaliates).
        """
        counter_chance = max(
            EVASION_COUNTER_MIN,
            EVASION_COUNTER_BASE - recursion_depth * EVASION_COUNTER_STEP,
        )
        return random.random() < counter_chance

File: genesis-web/src/core/combat/DiceResolver.ts
// Dice rolling and outcome resolution.

import {
  BOOSTED_MULTIPLIER,
  GRAZE_CHIP_MULTIPLIER,
  COUNTER_BASE,
  COUNTER_STEP,
  COUNTER_MIN,
} from '../constants'
import type { DiceProbabilities } from './HitChanceEvaluator'

/**
 * The four outcomes, unchanged in name and magnitude by the move to two phases.
 *
 * `Graze` was called `Fail` when it meant a clean miss. It has delivered chip
 * damage and an AP refund since, so the old name misdescribed it — and a name
 * that lies about behaviour is how the spec and the engine drifted apart in the
 * first place.
 */
export type DiceOutcome = 'Boosted' | 'Hit' | 'Evade' | 'Graze'

export interface OutcomeResult {
  output:  number   // final damage or healing value
  evaded:  boolean  // true when the attack was fully evaded
}

/**
 * Weighted random selection from any probability table.
 *
 * Generic over the band type so both phases roll through one implementation:
 * phase 1 picks a StrikeBand, phase 2 a ReactionBand, and the outcome is their
 * pairing. `fallback` is returned only if the weights never reach the roll,
 * which a table summing to 1 cannot do.
 */
export function rollTable<K extends string>(
  probs: Readonly<Record<K, number>>,
  fallback: K,
): K {
  const rand = Math.random()
  let cumulative = 0
  for (const [band, prob] of Object.entries(probs) as [K, number][]) {
    cumulative += prob
    if (rand < cumulative) return band
  }
  return fallback
}

/** The combined four-outcome table, for callers that already hold one. */
export function roll(probs: DiceProbabilities): DiceOutcome {
  return rollTable(probs, 'Hit')
}

/**
 * How much of a skill's output an outcome delivers.
 *
 * This is the single source of truth for outcome magnitude. It used to live
 * only inside applyOutcome, which nothing in the battle pipeline called — so
 * BOOSTED_MULTIPLIER never fired and a Boosted hit dealt exactly as much as a
 * plain Hit. The dice had four names but two behaviours.
 */
export function outcomeScale(outcome: DiceOutcome): number {
  switch (outcome) {
    case 'Boosted': return BOOSTED_MULTIPLIER
    case 'Hit':     return 1
    case 'Graze':   return GRAZE_CHIP_MULTIPLIER
    case 'Evade':   return 0
  }
}

// Apply a dice outcome to a raw output value and produce the full result.
export function applyOutcome(outcome: DiceOutcome, rawOutput: number): OutcomeResult {
  return {
    output: Math.round(rawOutput * outcomeScale(outcome)),
    evaded: outcome === 'Evade',
  }
}

// Counter chain — chance diminishes with each recursion depth.
export function resolveCounterRoll(depth: number): boolean {
  const chance = Math.max(
    COUNTER_MIN,
    COUNTER_BASE - depth * COUNTER_STEP,
  )
  return Math.random() < chance
}
